substrate/frame/nis/src/lib.rs

```diff
@@ -693,7 +702,7 @@ pub mod pallet {
 			let (owner, mut on_hold) = receipt.owner.ok_or(Error::<T>::AlreadyCommunal)?;
 			ensure!(owner == who, Error::<T>::NotOwner);
 
-			let now = frame_system::Pallet::<T>::block_number();
+			let now = T::BlockNumberProvider::current_block_number();
```

Comment:
Misses a migration for the `Receipts` to change the `expiry` to the "new clock". (Not required when switched to `System` as block number provider.

Reply:
Most of those PR is for pallet in the relay chain to go to a parachain without requiring migration.

But in the wider scope I can see parachain could also make use of the relay chain block number provider when they are going more agile and skip some blocks.
In this case a migration would help indeed. We can just provide a function helper to do this migration, no need for storage versioning. Because parachain could change the configuration of the block number whenever they want, so unrelated to the version of the pallet.

Reply:
I would not include a migration for the time being, assuming the PR and code-doc are super clear under which circumstances you do, and do not need a migration.

Reply:
> But in the wider scope I can see parachain could also make use of the relay chain block number provider when they are going more agile and skip some blocks.

This is the entire point of this exercise. Otherwise it is quite useless. Not sure why we approve prs without a migration. Also requiring everyone to write their own migration is a little bit stupid. Especially as by just "reading the docs" it will not be that easy. Block numbers are stored internally in some structures and may not be that obvious for others to find them etc. This will just end up in a mess.

Reply:
Should be converted to Instantiable pallet to test the migration on different mock runtimes using `System` and Relay BlockNumberProviders?

Reply:
> Misses a migration for the Receipts to change the expiry to the "new clock". (Not required when switched to System as block number provider.

Yeah, would need to type check the `BlockNumberProvider`?
```rust
#[cfg(feature = "try-runtime")]
fn pre_upgrade() -> Result<Vec<u8>, TryRuntimeError> {
    use sp_runtime::TryRuntimeError;

    if std::any::TypeId::of::<<T as Config>::BlockNumberProvider>() != std::any::TypeId::of:: 
   <frame_system::pallet_prelude::BlockNumberFor<T>>() {
        // If it's still using System, no migration needed.
        return Ok(Default::default())
    } else {
        return Err(TryRuntimeError::Other("No need to migrate"));
    }
}
```